keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,482 @@
+
+```proto
+message WatchRequest {
+  option (versionpb.etcd_version_msg) = "3.0";
+  // features that must be enabled in the cluster for the request to be accepted.
+  repeated string required_features = 4;
+  // request_union is a request to either create a new watcher or cancel an existing watcher.
+  oneof request_union {
+    WatchCreateRequest create_request = 1;
+    WatchCancelRequest cancel_request = 2;
+    WatchProgressRequest progress_request = 3 [(versionpb.etcd_version_field)="3.4"];
+  }
+}
+```
+If any of the `required_features` are not enabled in the cluster, the request will be immediately rejected, with response header
+
+```proto
+message ResponseHeader {
+  option (versionpb.etcd_version_msg) = "3.0";
+
+  // cluster_id is the ID of the cluster which sent the response.
+  uint64 cluster_id = 1;
+  // member_id is the ID of the member which sent the response.
+  uint64 member_id = 2;
+  // revision is the key-value store revision when the request was applied, and it's
+  // unset (so 0) in case of calls not interacting with key-value store.
+  // For watch progress responses, the header.revision indicates progress. All future events
+  // received in this stream are guaranteed to have a higher revision number than the
+  // header.revision number.
+  int64 revision = 3;
+  // raft_term is the raft term when the request was applied.
+  uint64 raft_term = 4;
+  // features in the request required_features that are not enabled in the cluster. 
+  repeated string missing_features = 4;
+}
+```
+
+### Feature Gate
+
+A feature can be registered as server level feature or cluster level feature, but not both.
+
+```go
+defaultServerFeatureGates := map[Feature]VersionedSpecs {
+		featureA: VersionedSpecs{
+			{Version: mustParseVersion("3.6"), Default: false, PreRelease: Beta},
+			{Version: mustParseVersion("3.7"), Default: true, PreRelease: GA},
+		},
+		featureB: VersionedSpecs{
+			{Version: mustParseVersion("3.7"), Default: false, PreRelease: Alpha},
+		},
+}
+defaultClusterFeatureGates := map[Feature]VersionedSpecs {
+		featureC: VersionedSpecs{
+			{Version: mustParseVersion("3.7"), Default: false, PreRelease: Beta},
+		},
+		featureD: VersionedSpecs{
+			{Version: mustParseVersion("3.6"), Default: false, PreRelease: Alpha},
+			{Version: mustParseVersion("3.7"), Default: true, PreRelease: Deprecated},
+		}
+}
+```
+
+The `func (s *EtcdServer) FeatureEnabled(key Feature) bool` interface would return if the feature is enabled for the whole cluster if it is registered as a cluster level feature.
+
+The feature gates for a server can only be set with the `--feature-gates`flag or in the `config-file` during startup. We do not support dynamically changing the feature gates when the server is running.
+
+### Cluster Level Feature Enablement
```

Comment:
Please discuss the process in a top down approach. I don't care about the proto details if I don't know how you plan to use it. Suggested order of topics:
* Properties that feature enablement needs
* How we plan to guarantee those properties
* Detail of the process.
* How proto will look.